heron/proto/ckptmgr.proto

```diff
@@ -0,0 +1,200 @@
+package heron.proto.ckptmgr;
+option java_package = "com.twitter.heron.proto.ckptmgr";
+option java_outer_classname = "CheckpointManager";
+
+import "common.proto";
+import "physical_plan.proto";
+
+// Message representing information about a globally consistent
+// checkpoint
+message StatefulConsistentCheckpoint {
+  required string checkpoint_id = 1;
+  // One can add more meta data about this ckpt later
+}
+
+// message stored in the state manager by the tmaster
+message StatefulConsistentCheckpoints {
+  // An ordered list of the globally consistent checkpoints
+  // that have been snapshotted and can be recovered from.
+  // 0th element is the most recent one, 1st is the next recent
+  // and so on
+  repeated StatefulConsistentCheckpoint consistent_checkpoints = 1;
+}
+
+// tmaster -> stmgr messages
+
+// Message sent to stmgrs by the tmaster to initiate checkpointing
+message StartStatefulCheckpoint {
+  required string checkpoint_id = 1;
+}
+
+// Message sent by tmaster to stmgr asking them to reset their instances
+// to this checkpoint
+message RestoreTopologyStateRequest {
+  required string checkpoint_id = 1;
+  // Every Attempt made by Tmaster to restore to a globally consistent
+  // checkpoint is tagged with a unique id. The stmgrs include this in their
+  // RestoreTopologyStateResponse below. This way tmaster could
+  // try restoring to the same checkpoint_id multiple times and still
+  // keep track of for which attempt the restore was in the responses
+  required int64 restore_txid = 2; 
+}
+
+// Message that stmgr sends to tmaster after it restores
+// all of its local instances to a checkpoint_id
+message RestoreTopologyStateResponse {
+  required heron.proto.system.Status status = 1;
+  required string checkpoint_id = 2;
+  required int64 restore_txid = 3; 
+}
+
+// Message sent by stmgr to tmaster asking it to reset the topology
+// to some valid checkpoint. This is sent either if stmgr dies
+// and comes back up or if an instance dies.
+message ResetTopologyState {
+  optional string dead_stmgr = 1;
+  optional int32 dead_taskid = 2;
+}
+
+// Message sent by stmgr to tmaster informing it about
```

Comment:
Why does the tmaster need to be aware of specific instance state? Why not have the stream manager manage the states of the instances under it's control?